ir/instr.cpp

```diff
@@ -824,6 +824,130 @@ bool BinOp::isDivOrRem() const {
 }
 
 
+vector<Value*> VectorPredicatedBinOp::operands() const {
+  return { lhs, rhs, mask, evl };
+}
+
+bool VectorPredicatedBinOp::propagatesPoison() const {
+  return true;
+}
+
+void VectorPredicatedBinOp::rauw(const Value &what, Value &with) {
+  RAUW(lhs);
+  RAUW(rhs);
+  RAUW(mask);
+  RAUW(evl);
+}
+
+void VectorPredicatedBinOp::print(ostream &os) const {
+  const char *str = nullptr;
+  switch (op) {
+  case Add:  str = "vp.add "; break;
+  case Sub:  str = "vp.sub "; break;
+  case Mul:  str = "vp.mul "; break;
+  case SDiv: str = "vp.sdiv "; break;
+  case UDiv: str = "vp.udiv "; break;
+  case SRem: str = "vp.srem "; break;
+  case URem: str = "vp.urem "; break;
+  case Shl:  str = "vp.shl "; break;
+  case AShr: str = "vp.ashr "; break;
+  case LShr: str = "vp.lshr "; break;
+  case And:  str = "vp.and "; break;
+  case Or:   str = "vp.or "; break;
+  case Xor:  str = "vp.xor "; break;
+  }
+
+  os << getName() << " = " << str << print_type(getType())
+     << lhs->getName()  << ", " << rhs->getName() << ", "
+     << mask->getName() << ", " << evl->getName();
+}
+
+StateValue VectorPredicatedBinOp::toSMT(State &s) const {
+  function<expr(const expr&, const expr&)> fn;
+
+  switch (op) {
+  case Add:  fn = [&](auto a, auto b) -> expr { return a + b; };     break;
+  case Sub:  fn = [&](auto a, auto b) -> expr { return a - b; };     break;
+  case Mul:  fn = [&](auto a, auto b) -> expr { return a * b; };     break;
+  case SDiv: fn = [&](auto a, auto b) -> expr { return a.sdiv(b); }; break;
+  case UDiv: fn = [&](auto a, auto b) -> expr { return a.udiv(b); }; break;
+  case SRem: fn = [&](auto a, auto b) -> expr { return a.srem(b); }; break;
+  case URem: fn = [&](auto a, auto b) -> expr { return a.urem(b); }; break;
+  case Shl:  fn = [&](auto a, auto b) -> expr { return a << b; };    break;
+  case AShr: fn = [&](auto a, auto b) -> expr { return a.ashr(b); }; break;
+  case LShr: fn = [&](auto a, auto b) -> expr { return a.lshr(b); }; break;
+  case And:  fn = [&](auto a, auto b) -> expr { return a & b; };     break;
+  case Or:   fn = [&](auto a, auto b) -> expr { return a | b; };     break;
+  case Xor:  fn = [&](auto a, auto b) -> expr { return a ^ b; };     break;
+  }
+
+  auto scalar_op = [&](auto a, auto ap, auto b, auto bp) -> StateValue {
+    auto v = fn(a, b);
+
+    if (isDivOrRem())
+      div_ub(s, a, b, ap, bp, op == SDiv || op == SRem);
```

Comment:
If it's really just a vector+select, then let's lower it like that in llvm2alive instead. There's no need for new ops in Alive2 IR.

Reply:
I don't think there exists a trivial lowering for %evl masking. That said, is it a good idea to lower those intrinsics to IR::BinOp+select, meanwhile we add an optional %evl field for IR::BinOp? We handle the %evl in IR::BinOp::toSMT. I think we will have good code reuse if we go this way.

Reply:
I am implementing the llvm.masked.load/store semantic which is similar to llvm.vp.*, the langref specifies two possible lowerings of those intrinsics https://llvm.org/docs/LangRef.html#id906

  >  It is useful for targets that support vector masked loads and allows vectorizing predicated basic blocks on these targets. Other targets may support this intrinsic differently, for example by lowering it into a sequence of branches that guard scalar load operations.

Lowering llvm.masked.load to "load + select" in llvm2alive won't work in this semantic, suppose all lanes are disabled by mask, there is no loads operation being executed in Langref's semantic, while in "load + select" the loads are always executed first, followed by masking. Similar problem applies to llvm.masked.store.

How should we encode the semantic for llvm.masked.load/store? A possible way would be creating a bunch of branches, just like what the Langref describes. Do you have any better ideas, @nunoplopes ?

Reply:
I would investigate first the semantics of the masked div, etc. If the execution of those is also predicated, then it may make sense to introduce a predicate vector instruction in Alive2 IR that executes element-wise instructions based on a mask or on a condition. Introducing branching will create a lot of phis and stuff, so not ideal.
We don't have API support for conditional stores. Some work in state/memory will be needed.